lib/Runtime/Library/JavascriptPromise.cpp

```diff
@@ -559,6 +559,147 @@ namespace Js
+        //4. Assert IsConstructor(C)
+        Assert(JavascriptOperators::IsConstructor(constructor));
+
+        //5. If IsCallable(onFinally) is false
+        //a. Let thenFinally be onFinally
+        //b. let catchFinally be onFinally 
+        //6. Else,
+        //  a. Let thenFinally be a new built-in function object as defined in ThenFinally Function.
+        //  b. Let catchFinally be a new built-in function object as defined in CatchFinally Function.
+        //  c. Set thenFinally and catchFinally's [[Constructor]] internal slots to C.
+        //  d. Set thenFinally and catchFinally's [[OnFinally]] internal slots to onFinally.
+
+        Var thenFinally;
+        Var catchFinally;
+
+        if (args.Info.Count > 1)
+        {
+            if (JavascriptConversion::IsCallable(args[1]))
+            {
+                //note to avoid duplicating code the ThenFinallyFunction works as both thenFinally and catchFinally using a flag
+                thenFinally = library->CreatePromiseThenFinallyFunction(EntryThenFinallyFunction, RecyclableObject::FromVar(args[1]), constructor, false);
+                catchFinally = library->CreatePromiseThenFinallyFunction(EntryThenFinallyFunction, RecyclableObject::FromVar(args[1]), constructor, true);
+            }
+            else
+            {
+                thenFinally = args[1];
+                catchFinally = args[1];
+            }
+        }
+        else
+        {
+            thenFinally = library->GetUndefined();
+            catchFinally = library->GetUndefined();
+        }
+
+        //7. Return ? Invoke(promise, "then", « thenFinally, catchFinally »).
+        Var funcVar = JavascriptOperators::GetProperty(promise, Js::PropertyIds::then, scriptContext);
+        if (!JavascriptConversion::IsCallable(funcVar))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_NeedFunction, _u("Promise.prototype.finally"));
+        }
+        RecyclableObject* func = RecyclableObject::UnsafeFromVar(funcVar);
+
+        return CALL_FUNCTION(scriptContext->GetThreadContext(),
+            func, Js::CallInfo(CallFlags_Value, 3),
+            promise,
+            thenFinally,
+            catchFinally);
+    }
+
+    //ThenFinallyFunction as described in draft of ES2018  section 25.4.5.3.1
+    //AND CatchFinallyFunction as described in draft of ES2018  section 25.4.5.3.2
+    Var JavascriptPromise::EntryThenFinallyFunction(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        ScriptContext* scriptContext = function->GetScriptContext();
+        
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+
+        JavascriptPromiseThenFinallyFunction* This = JavascriptPromiseThenFinallyFunction::FromVar(function);
```

Comment:
I'm not a big fan of naming this `JavascriptPromiseThenFinallyFunction*` as `This`. Can we call it `finallyFunction` or `thenFinallyFunction` or something?

Reply:
OK (sorry Javascript is my native language was trying to make it look like this.property)

Reply:
No worries, we just don't usually use this naming convention in our codebase. 👍